docs/docs/reference/contextual/derivation.md

```diff
@@ -1,362 +1,365 @@
 ```scala
-  def derived[T] given Mirror.Of[T] = ...
-```
-That is, the `derived` method takes an implicit parameter of (some subtype of) type `Mirror` that defines the shape of the deriving type `T` and it computes the typeclass implementation according
-to that shape. A given `Mirror` instance is generated automatically for
+  sealed trait Mirror {
 
- - case classes and objects,
- - enums and enum cases,
- - sealed traits or classes that have only case classes and case objects as children.
+    /** the type being mirrored */
+    type MirroredType
+ 
+    /** the type of the elements of the mirrored type */
+    type MirroedElemTypes
 
-<!---
-This is all a user of typeclass derivation has to know. The rest of this page contains information needed to be able to write a typeclass that can appear in a `derives` clause. In particular, it details the means provided for the implementation of data generic `derived` methods.
---->
+    /** The mirrored *-type */
+    type MirroredMonoType
 
-The description that follows gives a low-level way to define a type class.
+    /** The name of the type */
+    type MirroredLabel <: String
 
-### The Shape Type
+    /** The names of the elements of the type */
+    type MirroredElemLabels <: Tuple
+  }
 
-For every class with a `derives` clause, the compiler computes the shape of that class as a type. For example, here is the shape type for the `Tree[T]` enum:
+  object Mirror {
+    /** The Mirror for a product type */
+    trait Product extends Mirror {
 
-```scala
-Cases[(
-  Case[Branch[T], (Tree[T], Tree[T])],
-  Case[Leaf[T], T *: Unit]
-)]
+      /** Create a new instance of type `T` with elements taken from product `p`. */
+      def fromProduct(p: scala.Product): MirroredMonoType
+    }
+
+    trait Sum extends Mirror { self =>
+      /** The ordinal number of the case class of `x`. For enums, `ordinal(x) == x.ordinal` */
+      def ordinal(x: MirroredMonoType): Int
+    }
+  }
 ```
-Informally, this states that
 
-> The shape of a `Tree[T]` is one of two cases: Either a `Branch[T]` with two
-  elements of type `Tree[T]`, or a `Leaf[T]` with a single element of type `T`.
+Product types (ie. case classes and objects, and enum cases) have mirrors which are subtypes of `Mirror.Product`. Sum
+types (ie. sealed class or traits with product children, and enums) have mirrors which are subtypes of `Mirror.Sum`.
+
+For the `Tree` ADT from above the following `Mirror` instances will be automatically provided by the compiler,
 
-The type constructors `Cases` and `Case` come from the companion object of a class
-`scala.compiletime.Shape`, which is defined in the standard library as follows:
 ```scala
-sealed abstract class Shape
+// Mirror for Tree
+Mirror.Sum {
+  type MirroredType = Tree
+  type MirroredElemTypes[T] = (Branch[T], Leaf[T])
```

Comment:
I’m confused that `MirroredElemTypes` now takes a type parameter, even though in the definition of `Mirror` it is *-kinded.

Reply:
`MirroredElemTypes` is defined in subtypes of `Mirror` where the appropriate kind is known. The definition I showed as a member of `Mirror` is illustrative rather than actual ... I guess I should at least mention this at that point, but I didn't want to complicate things. One option would be to define these types in `Mirror` with an `AnyKind` bound ... do you think that would be preferable to deferring the definitions?

Reply:
> One option would be to define these types in `Mirror` with an `AnyKind` bound ... do you think that would be preferable to deferring the definitions?

Maybe it’s not needed. Your explanation made it clear. We can probably just add somewhere that the shown `Mirror` definition is not the actual one, but a simplified version.